windows/win_chocolatey.ps1

```diff
@@ -96,153 +83,261 @@ Else
+    if (-not $source)
+    {
+        $cmd += " -source https://chocolatey.org/api/v2/"
+    }
+    elseif ($source -eq "chocolatey")
+    {
+        $cmd += " -source https://chocolatey.org/api/v2/"
+    }
+    elseif (($source -eq "windowsfeatures") -or `
+        ($source -eq "webpi") -or ($source -eq "ruby"))
+    {
+        $cmd += " -source $source"
+    }
+    else
+    {
+        Throw "source is $source - must be one of chocolatey, ruby, webpi or windowsfeatures."
+    }
+
+    if ($version)
+    {
+        $cmd += " -version $version"
+    }
+
+    if ($force)
+    {
+        $cmd += " -force"
+    }
+
+    $results = invoke-expression $cmd
+
+    if ($LastExitCode -ne 0)
+    {
+        Set-Attr $result "choco_error_cmd" $cmd
+        Set-Attr $result "choco_error_log" "$results"
+        Throw "Error installing $package" 
+    }
+
+    if ("$results" -match ' upgraded (\d+)/\d+ package\(s\)\. ')
+    {
+        if ($matches[1] -gt 0)
         {
             $result.changed = $true
         }
-        Else
-        {
-            Fail-Json $result "WebPI install error: $webpi_install_result"
-        }
     }
 }
-$expression = $executable
-if ($state -eq "present")
-{
-    $expression += " install $package"
-}
-Elseif ($state -eq "absent")
-{
-    $expression += " uninstall $package"
-}
-if ($force)
+
+Function Choco-Install 
 {
-    if ($state -eq "present")
+    [CmdletBinding()]
+    
+    param(
+        [Parameter(Mandatory=$true, Position=1)]
+        [string]$package,
+        [Parameter(Mandatory=$false, Position=2)]
+        [string]$version,
+        [Parameter(Mandatory=$false, Position=3)]
+        [string]$source,
+        [Parameter(Mandatory=$false, Position=4)]
+        [bool]$force,
+        [Parameter(Mandatory=$false, Position=5)]
+        [bool]$upgrade
+    )
+
+    if (Choco-IsInstalled $package)
     {
-        $expression += " -force"
+        if ($upgrade)
+        {
+            Choco-Upgrade -package $package -version $version -source $source -force $force
+        }
+
+        return
     }
-}
-if ($version)
-{
-    $expression += " -version $version"
-}
-if ($source -eq "chocolatey")
-{
-    $expression += " -source https://chocolatey.org/api/v2/"
-}
-elseif (($source -eq "windowsfeatures") -or ($source -eq "webpi") -or ($source -eq "ruby"))
-{
-    $expression += " -source $source"
+
+    $cmd = "$executable install -y $package"
+
+    if (-not $source)
+    {
+        $cmd += " -source https://chocolatey.org/api/v2/"
+    }
+    elseif ($source -eq "chocolatey")
+    {
+        $cmd += " -source https://chocolatey.org/api/v2/"
+    }
+    elseif (($source -eq "windowsfeatures") -or `
+        ($source -eq "webpi") -or ($source -eq "ruby"))
+    {
+        $cmd += " -source $source"
+    }
+    else
+    {
+        Throw "source is $source - must be one of chocolatey, ruby, webpi or windowsfeatures."
```

Comment:
Same comment on sources as above. Would like to allow folks to use their own custom hosted sources, but also that some of those alternatives don't work yet.